Implementing drawPath delegate for DrawGlyphsRecorder

vitorroriz opened the pull request:
#### 8434dfba0479e828260c310732d53334beab4d69
<pre>
Implementing drawPath delegate for DrawGlyphsRecorder
<a href="https://bugs.webkit.org/show_bug.cgi?id=262223">https://bugs.webkit.org/show_bug.cgi?id=262223</a>
<a href="https://rdar.apple.com/115721319">rdar://115721319</a>

Reviewed by Said Abou-Hallawa.

After a CoreText update, certain multi-coloured glyphs will no longer
be drawn as a glyph (drawGlyphs) but as a path. However, up to now
WebKit hasn&apos;t implemented the drawPath delegate, so this was making
multi-coloured glyphs not to be drawn and thefefore, breaking some our
tests. This path implements the drawPath delegate.

Some tests were relying on implementation details of COLRv0 by CoreText.
Multi-coloured glyphs, like the glyph for unicode &quot;A&quot; in Ahem-COLR.ttf won&apos;t
be drawn as a glyph but as a path now. Therefore, certain things like
line-dashing and strokes aren&apos;t supported, which is what COLRv0 expects
and it is consistency with other browser engines implementations. For that
reason we are rebasilining such tests. It might be that other tests
on different platforms not tested here will also need rebaseline.

Note: For Apple Platforms, these changes are only for Sonoma, when CoreText was updated. So we
are keeping the former behavior for the former MAC OSes baselines.

* LayoutTests/fast/text/canvas-color-fonts/linedash-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-color-COLR-expected.html.
* LayoutTests/fast/text/canvas-color-fonts/stroke-color-COLR-expected.html:
* LayoutTests/fast/text/canvas-color-fonts/stroke-color-shadow-COLR-expected.html:
* LayoutTests/fast/text/canvas-color-fonts/stroke-color-shadow-ctm-COLR-expected.html:
* LayoutTests/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html:
* LayoutTests/fast/text/canvas-color-fonts/stroke-gradient-COLR.html:
* LayoutTests/fast/text/glyph-display-lists/glyph-display-list-colr-unshared.html:
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-monterey/fast/text/canvas-color-fonts/linedash-COLR-expected-mismatch.html: Copied from LayoutTests/fast/text/canvas-color-fonts/linedash-COLR-expected-mismatch.html.
* LayoutTests/platform/mac-monterey/fast/text/canvas-color-fonts/stroke-color-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-color-COLR-expected.html.
* LayoutTests/platform/mac-monterey/fast/text/canvas-color-fonts/stroke-color-shadow-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-color-shadow-COLR-expected.html.
* LayoutTests/platform/mac-monterey/fast/text/canvas-color-fonts/stroke-color-shadow-ctm-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-color-shadow-ctm-COLR-expected.html.
* LayoutTests/platform/mac-monterey/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html.
* LayoutTests/platform/mac-ventura/fast/text/canvas-color-fonts/linedash-COLR-expected-mismatch.html: Renamed from LayoutTests/fast/text/canvas-color-fonts/linedash-COLR-expected-mismatch.html.
* LayoutTests/platform/mac-ventura/fast/text/canvas-color-fonts/stroke-color-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-color-COLR-expected.html.
* LayoutTests/platform/mac-ventura/fast/text/canvas-color-fonts/stroke-color-shadow-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-color-shadow-COLR-expected.html.
* LayoutTests/platform/mac-ventura/fast/text/canvas-color-fonts/stroke-color-shadow-ctm-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-color-shadow-ctm-COLR-expected.html.
* LayoutTests/platform/mac-ventura/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html.
 * LayoutTests/platform/mac-monterey/fast/text/resources/Ahem-COLR.ttf: Added.
 * LayoutTests/platform/mac-ventura/fast/text/resources/Ahem-COLR.ttf: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebCore/platform/graphics/DrawGlyphsRecorder.h:
* Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorderCoreText.cpp:
(WebCore::drawPath):
(WebCore::DrawGlyphsRecorder::createInternalContext):
(WebCore::DrawGlyphsRecorder::recordDrawPath):

 * LayoutTests/platform/gtk/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html: Added.
 * LayoutTests/platform/gtk/fast/text/resources/Ahem-COLR.ttf: Added.
 * LayoutTests/platform/wpe/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html: Added.
 * LayoutTests/platform/wpe/fast/text/resources/Ahem-COLR.ttf: Added.

Canonical link: <a href="https://commits.webkit.org/270167@main">https://commits.webkit.org/270167@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5652961550cadc43722765b233ccff0abd87a54d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24767 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3312 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26021 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26886 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22736 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25036 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/4991 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/749 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23081 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25012 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2361 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21382 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27469 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2066 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22313 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/28469 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22591 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22658 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26280 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2003 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/311 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3293 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/22063 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5925 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2454 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2354 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorderCoreText.cpp:
```diff
@@ -71,13 +72,21 @@ static CGError drawImage(CGContextDelegateRef delegate, CGRenderingStateRef rsta
 
+static CGError drawPath(CGContextDelegateRef delegate, CGRenderingStateRef rstate, CGGStateRef gstate, CGPathDrawingMode drawingMode, CGPathRef path)
+{
+    DrawGlyphsRecorder& recorder = *static_cast<DrawGlyphsRecorder*>(CGContextDelegateGetInfo(delegate));
```
Should we check for `nullptr` before dereferencing ?

shallawa commented on LayoutTests/fast/text/canvas-color-fonts/linedash-COLR-expected-mismatch.html:
```diff
@@ -10,17 +10,17 @@
 document.fonts.load(font).then(function() {
     let canvas = document.getElementById("canvas");
-    let context = canvas.getContext("2d");
+    let context = canvas.getContext("2d")
```
Why semicolon was removed?

shallawa commented on LayoutTests/fast/text/canvas-color-fonts/stroke-color-COLR-expected.html:
```diff
@@ -10,16 +10,24 @@
+    context.lineWidth = 2;
+    context.strokeStyle = "green";
+    context.strokeRect(100, 20, 100, 100);
+    context.strokeText("A", 200, 100);
```
The current expected page does not rely on `strokeText()`. So if it does not work correctly, the test will fail because the expected page uses `strokeRect()`. With your change we are comparing `strokeText()` with `strokeText()`. So if it does not work correctly, the test will not fail because the expected page will render the letter 'A' incorrectly also.

shallawa commented on LayoutTests/fast/text/canvas-color-fonts/stroke-color-shadow-COLR.html:
```diff
@@ -25,7 +25,7 @@
     context.shadowOffsetX = 100;
     context.shadowOffsetY = 105;
-    context.strokeText("BAB", 100, 100);
+    context.strokeText("B", 100, 100);
```
Now you are taking a different approach form what you did in the previous test. You are dropping the letter 'A' form the test.

shallawa commented on LayoutTests/fast/text/canvas-color-fonts/stroke-color-shadow-ctm-COLR.html:
```diff
@@ -25,7 +25,7 @@
     context.shadowOffsetX = 100;
     context.shadowOffsetY = 105;
-    context.strokeText("BAB", 100, 100);
+    context.strokeText("B", 100, 100);
```
Ditto.

shallawa commented on LayoutTests/fast/text/canvas-color-fonts/stroke-gradient-COLR.html:
```diff
@@ -25,7 +25,7 @@
     gradient.addColorStop(1, "rgb(255, 153, 51)");
     context.strokeStyle = gradient;
-    context.strokeText("BAB", 100, 100);
+    context.strokeText("BBB", 100, 100);
```
A third approach, you are replacing 'A' with 'B'.

shallawa commented on LayoutTests/fast/text/glyph-display-lists/glyph-display-list-colr-unshared.html:
```diff
@@ -17,8 +17,9 @@
 <div id=container>
-    <span id=first>A</span>
-    <span id=second>A</span>
+    <!-- This is only true for glyphs drawn as glyphs (like B for this font), not for glyphs drawn as paths (like A for this font)  -->
```
I do not understand this comment. What do you mean by "this" in "This is only true for ..."? Also this test is enabled for other ports. So this comment may be relevant to Apple ports only.

shallawa commented on LayoutTests/fast/text/glyph-display-lists/glyph-display-list-colr-unshared.html:
```diff
@@ -17,8 +17,9 @@
-    <span id=second>A</span>
+    <!-- This is only true for glyphs drawn as glyphs (like B for this font), not for glyphs drawn as paths (like A for this font)  -->
+    <span id=first>B</span>
+    <span id=second>B</span>
```
I think we need to find a way to test the glyphs which are drawn as paths in this font without using strokeText().

vitorroriz replied:
rdar://117866670

shallawa commented on Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorderCoreText.cpp:
```diff
@@ -379,6 +388,42 @@ void DrawGlyphsRecorder::recordDrawImage(CGRenderingStateRef, CGGStateRef gstate
+    auto gstateTransform = m_owner.getCTM();
+    // The path we get has already CTM applied to it but we should serailize the non-transformed version to correctly apply line width.
+    CGAffineTransform invertTransform = CGAffineTransformInvert(gstateTransform);
+    CGPathRef localPath = CGPathCreateCopyByTransformingPath(coreGraphicsPath, &invertTransform);
```
I think this will leak the `CGPathRef`. Use `adoptCF()`.

shallawa commented on Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorderCoreText.cpp:
```diff
@@ -379,6 +388,42 @@ void DrawGlyphsRecorder::recordDrawImage(CGRenderingStateRef, CGGStateRef gstate
+    updateCTM(*CGGStateGetCTM(gstate));
+    auto gstateTransform = m_owner.getCTM();
+    // The path we get has already CTM applied to it but we should serailize the non-transformed version to correctly apply line width.
+    CGAffineTransform invertTransform = CGAffineTransformInvert(gstateTransform);
```
`updateCTM()` sets the owner CTM to `CGGStateGetCTM()`. I think we can write this code like this:
```
auto* ctm = CGGStateGetCTM(gstate);
updateCTM(*ctm);
auto invertTransform = CGAffineTransformInvert(*ctm);
```

shallawa commented on LayoutTests/fast/text/canvas-color-fonts/linedash-COLR-expected.html:
```diff
@@ -0,0 +1,24 @@
+    font-family: "WebFont";
+    src: url("../resources/Ahem-COLR.ttf") format("truetype");
+}
+</style>
```
I think this font is not used.

shallawa commented on LayoutTests/fast/text/canvas-color-fonts/linedash-COLR-expected.html:
```diff
@@ -0,0 +1,24 @@
+<script>
+let canvas = document.getElementById("canvas");
+let context = canvas.getContext("2d");
+context.lineWidth = 2;
```
This test does not stroke anything. So why do we set the `lineWidth`?

shallawa commented on LayoutTests/fast/text/canvas-color-fonts/linedash-COLR-expected.html:
```diff
@@ -0,0 +1,24 @@
+<!DOCTYPE html>
```
Since you are adding the expected page, I think we should remove `linedash-COLR-expected-mismatch.html`.

vitorroriz replied:
I've just add the -mismatch files for Monterey/Ventura because the behavior changed with the CoreText update on Sonoma.

shallawa commented on LayoutTests/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html:
```diff
@@ -4,5 +4,22 @@
+let gradient = context.createLinearGradient(100, 200, 400, 200);
+gradient.addColorStop(0, "rgb(255, 0, 128)");
+gradient.addColorStop(1, "rgb(255, 153, 51)");
+context.strokeStyle = gradient;
```
I think this `strokeStyle` is not needed.

shallawa commented on LayoutTests/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html:
```diff
@@ -4,5 +4,22 @@
 <meta charset="utf-8">
 </head>
 <body>
```
I wonder why we checked in this empty expected page.

shallawa commented on LayoutTests/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html:
```diff
@@ -4,5 +4,22 @@
+gradient.addColorStop(0, "rgb(255, 0, 128)");
+gradient.addColorStop(1, "rgb(255, 153, 51)");
+context.strokeStyle = gradient;
+context.fillStyle = gradient;
+context.fillRect(200, 20, 100, 100);
+context.strokeStyle = "white";
+context.strokeRect(100, 20, 100, 100);
+context.strokeRect(200, 20, 100, 100);
+context.strokeRect(220, 20, 20, 100);
+context.strokeRect(300, 20, 100, 100);
```
I am not sure what we expect from the test here. The test draws "BAB" and then cleans the two 'B's leaving only the 'A'. Then it cleans the antialiased pixels in the middle of the 'A'.

So why do not we draw the 'A' only and maybe add pixel tolerance if it is needed?

vitorroriz replied:
When the test strokes the white "B"s it cuts the edges of "A". Also, I wanted to replicate the test the closest as possible to the steps it takes when drawing.

shallawa commented on LayoutTests/fast/text/glyph-display-lists/glyph-display-list-colr-unshared.html:
```diff
@@ -17,8 +17,8 @@
 <div id=container>
-    <span id=first>A</span>
-    <span id=second>A</span>
+    <span id=first>B</span>
+    <span id=second>B</span>
```
If we keep using 'A', will the display list be shared with your change? In other words, what is the difference between caching a display list for letter 'A' and caching a display list for letter 'B'?

shallawa commented on LayoutTests/platform/mac-monterey/fast/text/canvas-color-fonts/stroke-color-COLR-expected.html:
```diff
@@ -0,0 +1,25 @@
+<html>
+<head>
+<meta charset="utf-8">
+<style>
+@font-face {
+    font-family: "WebFont";
+    src: url("../resources/Ahem-COLR.ttf") format("truetype");
+}
+</style>
```
I think this can be deleted.

shallawa commented on Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorderCoreText.cpp:
```diff
@@ -379,6 +388,44 @@ void DrawGlyphsRecorder::recordDrawImage(CGRenderingStateRef, CGGStateRef gstate
+    // The path we get has already CTM applied to it but we should serialize the non-transformed version to correctly apply line width.
+    CGAffineTransform invertTransform = CGAffineTransformInvert(*ctm);
+    auto localPath = adoptCF(CGPathCreateCopyByTransformingPath(coreGraphicsPath, &invertTransform));
+    Path path { PathCG::create(adoptCF(CGPathCreateMutableCopy(localPath.get()))) };
```
I think `CGPathCreateMutableCopyByTransformingPath()` can be used instead.
```
auto localPath = adoptCF(CGPathCreateMutableCopyByTransformingPath(coreGraphicsPath, &invertTransform));
Path path { PathCG::create(WTFMove(localPath)) };
```

shallawa commented on Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorderCoreText.cpp:
```diff
@@ -379,6 +388,44 @@ void DrawGlyphsRecorder::recordDrawImage(CGRenderingStateRef, CGGStateRef gstate
+    auto localPath = adoptCF(CGPathCreateCopyByTransformingPath(coreGraphicsPath, &invertTransform));
+    Path path { PathCG::create(adoptCF(CGPathCreateMutableCopy(localPath.get()))) };
+
+    updateFillColor(CGGStateGetFillColor(gstate));
```
Should not we set the fillColor only when we fill or draw the path?

shallawa commented on Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorderCoreText.cpp:
```diff
@@ -379,6 +388,44 @@ void DrawGlyphsRecorder::recordDrawImage(CGRenderingStateRef, CGGStateRef gstate
+    Path path { PathCG::create(adoptCF(CGPathCreateMutableCopy(localPath.get()))) };
+
+    updateFillColor(CGGStateGetFillColor(gstate));
+    updateStrokeColor(CGGStateGetStrokeColor(gstate));
```
Should not we set the strokeColor only when we stroke or draw the path?

vitorroriz committed:
Implementing drawPath delegate for DrawGlyphsRecorder
https://bugs.webkit.org/show_bug.cgi?id=262223
rdar://115721319

Reviewed by Said Abou-Hallawa.

After a CoreText update, certain multi-coloured glyphs will no longer
be drawn as a glyph (drawGlyphs) but as a path. However, up to now
WebKit hasn't implemented the drawPath delegate, so this was making
multi-coloured glyphs not to be drawn and thefefore, breaking some our
tests. This path implements the drawPath delegate.

Some tests were relying on implementation details of COLRv0 by CoreText.
Multi-coloured glyphs, like the glyph for unicode "A" in Ahem-COLR.ttf won't
be drawn as a glyph but as a path now. Therefore, certain things like
line-dashing and strokes aren't supported, which is what COLRv0 expects
and it is consistency with other browser engines implementations. For that
reason we are rebasilining such tests. It might be that other tests
on different platforms not tested here will also need rebaseline.

Note: For Apple Platforms, these changes are only for Sonoma, when CoreText was updated. So we
are keeping the former behavior for the former MAC OSes baselines.

* LayoutTests/fast/text/canvas-color-fonts/linedash-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-color-COLR-expected.html.
* LayoutTests/fast/text/canvas-color-fonts/stroke-color-COLR-expected.html:
* LayoutTests/fast/text/canvas-color-fonts/stroke-color-shadow-COLR-expected.html:
* LayoutTests/fast/text/canvas-color-fonts/stroke-color-shadow-ctm-COLR-expected.html:
* LayoutTests/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html:
* LayoutTests/fast/text/canvas-color-fonts/stroke-gradient-COLR.html:
* LayoutTests/fast/text/glyph-display-lists/glyph-display-list-colr-unshared.html:
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-monterey/fast/text/canvas-color-fonts/linedash-COLR-expected-mismatch.html: Copied from LayoutTests/fast/text/canvas-color-fonts/linedash-COLR-expected-mismatch.html.
* LayoutTests/platform/mac-monterey/fast/text/canvas-color-fonts/stroke-color-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-color-COLR-expected.html.
* LayoutTests/platform/mac-monterey/fast/text/canvas-color-fonts/stroke-color-shadow-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-color-shadow-COLR-expected.html.
* LayoutTests/platform/mac-monterey/fast/text/canvas-color-fonts/stroke-color-shadow-ctm-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-color-shadow-ctm-COLR-expected.html.
* LayoutTests/platform/mac-monterey/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html.
* LayoutTests/platform/mac-ventura/fast/text/canvas-color-fonts/linedash-COLR-expected-mismatch.html: Renamed from LayoutTests/fast/text/canvas-color-fonts/linedash-COLR-expected-mismatch.html.
* LayoutTests/platform/mac-ventura/fast/text/canvas-color-fonts/stroke-color-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-color-COLR-expected.html.
* LayoutTests/platform/mac-ventura/fast/text/canvas-color-fonts/stroke-color-shadow-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-color-shadow-COLR-expected.html.
* LayoutTests/platform/mac-ventura/fast/text/canvas-color-fonts/stroke-color-shadow-ctm-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-color-shadow-ctm-COLR-expected.html.
* LayoutTests/platform/mac-ventura/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html: Copied from LayoutTests/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html.
 * LayoutTests/platform/mac-monterey/fast/text/resources/Ahem-COLR.ttf: Added.
 * LayoutTests/platform/mac-ventura/fast/text/resources/Ahem-COLR.ttf: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebCore/platform/graphics/DrawGlyphsRecorder.h:
* Source/WebCore/platform/graphics/coretext/DrawGlyphsRecorderCoreText.cpp:
(WebCore::drawPath):
(WebCore::DrawGlyphsRecorder::createInternalContext):
(WebCore::DrawGlyphsRecorder::recordDrawPath):

 * LayoutTests/platform/gtk/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html: Added.
 * LayoutTests/platform/gtk/fast/text/resources/Ahem-COLR.ttf: Added.
 * LayoutTests/platform/wpe/fast/text/canvas-color-fonts/stroke-gradient-COLR-expected.html: Added.
 * LayoutTests/platform/wpe/fast/text/resources/Ahem-COLR.ttf: Added.

Canonical link: https://commits.webkit.org/270167@main
